src/offset_of.rs

```diff
@@ -44,33 +44,46 @@
 #[cfg(memoffset_maybe_uninit)]
 macro_rules! offset_of {
     ($parent:tt, $field:tt) => (unsafe {
+        // Make sure the field actually exists. This line ensures that a
+        // compile-time error is generated if $field is accessed through a
+        // Deref impl.
+        let $parent { $field: _, .. };
+
         // Create an instance of the container and calculate the offset to its field.
         // Here we're using an uninitialized instance of $parent.
-        // Since we're not using its field, there's no UB caused by reading uninitialized memoty.
+        // Since we're not using its field, there's no UB caused by reading uninitialized memory.
         // There *IS*, though, UB caused by creating references to uninitialized data,
         // which is illegal since the compiler is allowed to assume that a reference
         // points to valid data.
-        // AFAIK we cannot avoid UB completely.
+        // However, currently we just cannot avoid UB completely.
         let val = $crate::mem::MaybeUninit::<$parent>::uninit();
-        let &$parent { $field: ref f, .. } = &*val.as_ptr();
-        (f as *const _ as *const u8 as usize) - (val.as_ptr() as *const u8 as usize)
+        let base_ptr = val.as_ptr();
+        #[allow(unused_unsafe)] // for packed structs
+        let field_ptr = unsafe { &(*base_ptr).$field as *const _ };
+        let offset = (field_ptr as usize) - (base_ptr as usize);
+        offset
```

Comment:
The last 4 lines are now the same for both versions of the macro. Is there any good way to share them?

Reply:
Wait, if the pattern destructuring is not necessary, we can support nested member access.

Reply:
Look up a few lines. The check is still there. Just done differently.